librz/asm/p/asm_snes.c

```diff
@@ -49,6 +34,19 @@ static int dis(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
 	return dlen;
 }
 
+static bool snes_asm_init(void **user) {
+	*user = RZ_NEW0(struct snes_asm_flags);
+	return user != NULL;
+}
+
+static bool snes_asm_fini(void *user) {
+	if (!user) {
```

Comment:
I wonder if it makes sense to use `rz_return_val_if_fail(user, false)` to catch the programmer errors, since this shouldn't happen, I suppose?

Reply:
> > > I haven't look at this in detail yet, but I think we should pay particular attention to the fact that this PR calls init/fini of each plugin every time you change the currently selected plugin. I wonder whether it is the right thing to do from a concept point of view and from the implementation point of view.
> > 
> > 
> > This is ok. there are only a bunch of archs that does this and it is the correct way of initializing the stuff, so people can actually use the library multiple times with different contexts
> 
> The reason I said that is because of this scenario:
> 
> ```
> [0x00000000]> wx 00b5884206bf01300130042000bd
> [0x00000000]> e asm.arch=arm asm.bits=16
> [0x00000000]> pd 7
>             0x00000000      push  {lr}
>             0x00000002      cmp   r0, r1
>             0x00000004      itte  eq
>             0x00000006      addeq r0, 1
>             0x00000008      addeq r0, 1
>             0x0000000a      movne r0, 4
>             0x0000000c      pop   {pc}
> [0x00000000]> pd 2 @ 6
>             0x00000006      addeq r0, 1
>             0x00000008      addeq r0, 1
> [0x00000000]> pd 2 @ 6 @a:x86
>             0000:0006      add   word [bx + si], si
>             0000:0008      add   word [bx + si], si
> [0x00000000]> pd 2 @ 6
>             0x00000006      adds  r0, 1
>             0x00000008      adds  r0, 1
> [0x00000000]> 
> ```
> 
> Some instructions in ARM depends on previous ones (in particular, the `it` instruction has some suffixes that changes how to interpret the following X instructions) and the arm plugin has a state to keep track of how to interpret them. As you can see, changing temporarily the plugin to x86 makes the arm plugin lose the state, so it doesn't know anymore how to interpret those instructions.
> 
> That said, it should not be a big problem and we can probably ignore it because anyway even right now, if you just go directly to address `6`, the asm plugin won't know how to interpret the adds instruction (because he never saw the `itte` instruction).
> 
> Just wanted to raise this.

I think this is correct behaviour: if you change ASM you need to change also the context. makes 0 sense to me to keep that context.

Reply:
Why was this marked as resolved?